### PR TITLE
Apply Tim Ruffin RFC review on aggregation speed-up

### PR DESCRIPTION
**Pull request by Nashtare**
This PR implements an optimization of the aggregate() method, that can yield up to 90% speed-up for this phase when running with large settings (i.e. 666-out-of-1000 for instance), and still yields nice reduction factors with smaller instances (3 times faster on my machine for 66-out-of-100).

It consists in computing the aggregated signature first, without verifying individual signatures, and verifying solely the aggregated signature. If the verification fails, then we proceed to individual signatures checks to identify the malicious participants, as was done previously.

~~It induces an API change though (the `aggregate()` method now takes the `message_hash` as argument). @atanmarko as you're working on the FROST integration, I'm tagging you as well in case this requires some coordination on the topos-node side.~~

Ref: [https://mailarchive.ietf.org/arch/msg/cfrg/QQhyjvvcoaqLslaX3gWwABqHN-s/](https://mailarchive.ietf.org/arch/msg/cfrg/QQhyjvvcoaqLslaX3gWwABqHN-s/)

### CODE REVIEW

File: src/signature.rs [hunk 1]
```diff
@@ -723,58 +726,68 @@ impl SignatureAggregator<Finalized> {
             .collect();
         let mut z = Scalar::zero();
 
+        // We first combine all partial signatures together, to remove the need for individual
+        // signature verification in case the final group signature is valid.
         for signer in self.state.signers.iter() {
-            // [DIFFERENT_TO_PAPER] We're not just pulling lambda out of our
-            // ass, instead to get the correct algebraic properties to allow for
```
Comment (4l0n50) [inline src/signature.rs:728]: Lol, it says "ass"

File: src/signature.rs [hunk 2]
```diff
@@ -706,7 +706,10 @@ impl SignatureAggregator<Finalized> {
     /// A Result whose Ok() value is a [`ThresholdSignature`], otherwise a
     /// `BTreeMap<u32, &'static str>` containing the participant indices of the misbehaving
     /// signers and a description of their misbehaviour.
-    pub fn aggregate(&self) -> Result<ThresholdSignature, BTreeMap<u32, &'static str>> {
+    pub fn aggregate(
+        &self,
+        message_hash: &[u8; 32],
+    ) -> Result<ThresholdSignature, BTreeMap<u32, &'static str>> {
```
Comment (4l0n50) [inline src/signature.rs:712]: Why do you need `message_hash`now?

**Review by 4l0n50 — APPROVED** [review]
Thir PR adds an "optimistic verification", which avoids checking correctness of individual shares when the final signature is correct.

File: src/signature.rs (hunk 2 above)
Comment (Nashtare) [reply src/signature.rs:712]: We compute the final signature verification, which endures that the individual signatures were correct. This requires the message hash.

Comment (4l0n50) [reply src/signature.rs:712]: but why it was not used before? you need to hash the message in order to check correctness of the shares

Comment (Nashtare) [reply src/signature.rs:712]: Ah yeah, I thought the challenge was already known by the aggregator. The message is actually part of its fields, so we can retrieve it without changing the API. Good point, thanks!

Comment (Nashtare) [reply src/signature.rs:712]: Oh yeah, I thought the challenge was already known by the aggregator. The message is actually part of its fields, so we can retrieve it without changing the API. Good point, thanks!

Comment (Nashtare) [reply src/signature.rs]: I updated it. Would need you to review again (it dismisses automatically reviews when pushing changes)

**Review by 4l0n50 — APPROVED** [review]

**Pull request merged**